monai/data/dataset.py

```diff
@@ -30,33 +34,116 @@ class Dataset(torch.utils.data.Dataset):
+                transform = Compose([transform])
+            else:
+                transform = Compose(transform)
         self.transform = transform
+        self.cache_dir = Path(cache_dir) if cache_dir is not None else None
+
+    def _pre_first_random_transform(self, item_transformed):
+        """
+        Process the data from original state up to the first random element.
+
+        Args:
+            item_transformed: The data to be transformed
+        Returns:
+            the transformed element up to the first identified
+            random transform object
+        """
+        for _transform in self.transform.transforms:
+            # execute all the deterministic transforms before the first random transform
+            if isinstance(_transform, Randomizable):
+                break
+            item_transformed = apply_transform(_transform, item_transformed)
+        return item_transformed
+
+    def _first_random_and_beyond_transform(self, item_transformed):
+        """
+        Process the data from before the first random transform to the final state ready for evaluation.
+        Args:
+            item_transformed: The data to be transformed (already process upto the first random transform)
+        Returns:
+            the transformed element through the random transforms
+        """
+        start_post_randomize_run = False
+        for _transform in self.transform.transforms:
+            if start_post_randomize_run or isinstance(_transform, Randomizable):
+                start_post_randomize_run = True
+                item_transformed = apply_transform(_transform, item_transformed)
+        return item_transformed
+
+    def _pre_first_random_transformed_cached_instance(self, item_transformed):
+        """
+            A function to cache the expensive input data transform operations
+            so that huge data sets (larger than computer memory) can be processed
+            on the fly as needed, and intermediate results written to disk for
+            future use.
+        Args:
+            item_transformed: The current data element to be mutated into transformed representation
+
+        Returns:
+            The transformed data_element, either from cache, or explicitly computing it.
+
+        Warning:
+            The current implementation does not encode transform information as part of the
+            hashing mechanism used for generating cache names.  If the transforms applied are
+            changed in any way, the objects in the cache dir will be invalid.  The hash for the
+            cache is ONLY dependant on the input filename paths.
+        """
+        if item_transformed.get("cached", False) is False:
+            hashfile = None
+            if self.cache_dir is not None:
+                cache_dir_path: Path = Path(self.cache_dir)
+                if cache_dir_path.is_dir():
+                    # TODO: Find way to hash transforms content as part of the cache
```

Comment:
We discussed hash method internally before, it may not be a good idea, because:
1. The float data for hash on different devices(CPU/GPU, FP32/FP16) can be slightly different which will output different hash value.
2. If the data content is very big(maybe stream buffer instead of file), it takes time and resources to compute hash.

So I think maybe the easiest way is to change self._cache as `dict` type and save cached data with `index` as key(and cache file name maybe). For same dataset, the order of data list should be always the same.

Reply:
We are hashing strings of filenames, NOT tensors!  Once we have tensors, we do not compute the hash.  My understanding is that at this point the of the processing only dictionaries with paths to original images would be encountered.   It is not until AFTER the transforms are applied that the dictionaries have tensors in them.

We are mapping the input of "strings to files" to the outputs of "pre-computed tensors of all the non-random transforms".

Reply:
Sorry maybe I didn't make it clear, I mean the existing examples use `filenames` to load data, but if users provide a buffer, it will also be OK(coming feature will continuously load data from streaming buffer). Then you need to hash on real data.
For example, if the initial input data is from a `data buffer`(dict({'img': numpy.ndarray, ...})) instead `LoadNifti`
```py
train_transforms = Compose([
    AddChanneld(...),
    RandRotated(...),
    ... ...
    ToTensor(...)
])
```
Thanks.

Reply:
This solution probably won't work for buffers, but it probably would not be needed for that case. This solution was predicated on the assumption that inputs would be file names. It won't work for buffered data. Perhaps the use of the cache_dir can only be used in the very common case of filenames, and dis-allowed for any non-filestring inputs.

Reply:
Yes, I agree that in most cases, we just need to load Dataset for files.
There is another potential problem is that:
If the input data contains more extra data except 'image' and 'label', it may contain float number, which is necessary for some tasks. It may affect the hash operation?
For example:
```py
{
    'image': './train_image1.nii.gz',
    'label': './train_label1.nii.gz',
    'some weights': np.ndarray(XXX)
}

Reply:
This would cause a problem if 'some weights' changes during a single run, and may further complicate the re-usability of the persistence for different runs, but within a single run we can achieve performance improvements that considerably improved over the simple Dataset processing. 

I think this is a reasonable first step that can greatly speed up processing of huge data sets.